packages/tangle/filter.go

```diff
@@ -234,3 +244,67 @@ func (f *RecentlySeenBytesFilter) getRejectCallback() (result func(bytes []byte,
 	f.onRejectCallbackMutex.Unlock()
 	return
 }
+
+// NewTransactionFilter creates a new transaction filter.
+func NewTransactionFilter() *TransactionFilter {
+	return &TransactionFilter{}
+}
+
+// TransactionFilter filters messages based on their timestamps and transaction timestamp.
+type TransactionFilter struct {
+	onAcceptCallback func(msg *Message, peer *peer.Peer)
+	onRejectCallback func(msg *Message, err error, peer *peer.Peer)
+
+	onAcceptCallbackMutex sync.RWMutex
+	onRejectCallbackMutex sync.RWMutex
+}
+
+// Filter compares the timestamps between the message and it's transaction payload and calls the corresponding callback.
+func (f *TransactionFilter) Filter(msg *Message, peer *peer.Peer) {
+	if payload := msg.Payload(); payload.Type() == ledgerstate.TransactionType {
+		transaction, _, err := ledgerstate.TransactionFromBytes(payload.Bytes())
+		if err != nil {
+			f.getRejectCallback()(msg, err, peer)
+			return
+		}
+		if !isMessageAndTransactionTimestampsValid(transaction, msg) {
+			f.getRejectCallback()(msg, ErrInvalidMessageAndTransactionTimestamp, peer)
+			return
+		}
+	}
+	f.getAcceptCallback()(msg, peer)
+}
+
+func isMessageAndTransactionTimestampsValid(transaction *ledgerstate.Transaction, message *Message) bool {
+	transactionTimestamp := transaction.Essence().Timestamp()
+	messageTimestamp := message.IssuingTime()
+	return transactionTimestamp.Before(messageTimestamp) && transactionTimestamp.Sub(messageTimestamp) <= MaxReattachmentTimeMin
```

Comment:
2 things here:
- the first part of the check  `transactionTimestamp.Before(messageTimestamp)` will return false if the times are equal. Unfortunately, we recently realised that on Windows the precision of time.Now() is in the order of millisecond, thus it could happen, especially during testing, that those times are the same.
- the second part of the check has a problem: imagine that transactionTimestamp is 0, and messageTimestamp is 100 => than you have 0-100 = -100 <= MaxReattachmentTime

Reply:
I guess that having a unit test for these kind of functions always help